### PR TITLE
Non-parameterized versions of Linear and Convolution2D

### LINKED ISSUE #9: Non-parameterized versions of Linear and Convolution2D
Current implementation of `Linear` and `Convolution2D` use their own parameter arrays. It lacks flexibility of the architecture design. Non-parameterized Linear is a fully-connected layer that uses some of input variables as its *parameter*. It should look like:

```
x, W, b = Variable(...), Variable(...), Variable(...)
y = F.linear(x, W=W, b=b)
```

Non-parameterized Convolution2D should have similar interface.

### PR DESCRIPTION
**Pull request by okuta**
fixes #9

### CODE REVIEW

File: chainer/functions/nonparameterized_convolution_2d.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            ``stride=s`` and ``stride=(s, s)`` are equivalent.
+        pad (int or (int, int)): Spatial padding width for input arrays.
+            ``pad=p`` and ``pad=(p, p)`` are equivalent.
+        use_cudnn (bool): If True, then this function uses CuDNN if available.
+
+    .. seealso:: :meth:`Convolution2D`
+
+    """
+    def __init__(self, stride=1, pad=0, use_cudnn=True):
+        self.stride = stride
```
Comment (unnonouno) [inline chainer/functions/nonparameterized_convolution_2d.py]: Do we need to check type of `stride`?

**Comment by unnonouno** [comment]
How about use `initialW` in the constructors to set initial parameters?

**Comment by unnonouno** [comment]
Please rebase or merge master branch.

**Comment by okuta** [comment]
Thank you for your comments.
I added arguments check.

File: chainer/functions/nonparameterized_convolution_2d.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        W = x[1]
+        b = None
+        if len(x) == 3:
+            b = x[2]
+        func = conv2d_module.Convolution2D(
+            W.shape[1], W.shape[0], W.shape[2:],
+            stride=self.stride, pad=self.pad, use_cudnn=self.use_cudnn,
+            initialW=W, initial_bias=b)
+        self.func = func
+        if any(isinstance(i, cuda.GPUArray) for i in x):
```
Comment (unnonouno) [inline chainer/functions/nonparameterized_convolution_2d.py]: why not to use `func.forward(x[:1])`?

File: chainer/functions/nonparameterized_convolution_2d.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            func.to_gpu()
+            return func.forward_gpu(x[:1])
+        else:
+            return func.forward_cpu(x[:1])
+
+    def backward_cpu(self, x, gy):
```
Comment (unnonouno) [inline chainer/functions/nonparameterized_convolution_2d.py]: Are both `backward_cpu` and `backward_gpu` are required? How about calling `func.backward`?

File: chainer/functions/nonparameterized_linear.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    Args:
+        x (~chainer.Variable): Input variable.
+        W (~chainer.Variable): Weight variable.
+        b (~chainer.Variable): Bias variable.
+
+    Returns:
+        ~chainer.Variable: Output variable.
+
+    .. seealso:: :meth:`Linear`
```
Comment (unnonouno) [inline chainer/functions/nonparameterized_linear.py]: class?

File: chainer/functions/nonparameterized_convolution_2d.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        stride (int or (int, int)): Stride of filter applications.
+            ``stride=s`` and ``stride=(s, s)`` are equivalent.
+        pad (int or (int, int)): Spatial padding width for input arrays.
+            ``pad=p`` and ``pad=(p, p)`` are equivalent.
+        use_cudnn (bool): If True, then this function uses CuDNN if available.
+
+    Returns:
+        ~chainer.Variable: Output variable.
+
+    .. seealso:: :meth:`Convolution2D`
```
Comment (unnonouno) [inline chainer/functions/nonparameterized_convolution_2d.py]: class?

File: chainer/functions/linear.py [hunk 6]
```diff
@@ -38,6 +38,8 @@ class Linear(function.Function):
         wscale (float): Scaling factor of the weight matrix.
         bias (float): Initial bias value.
         nobias (bool): If True, then this function does not use the bias.
+        initialW (2-D array): Initial weight value.
```
Comment (unnonouno) [inline chainer/functions/linear.py]: Please write behavior when `initialW == None`

File: chainer/functions/nonparameterized_convolution_2d.py (hunk 3 above)
Comment (okuta) [inline chainer/functions/nonparameterized_convolution_2d.py]: I fiexed.

File: chainer/functions/nonparameterized_convolution_2d.py (hunk 2 above)
Comment (okuta) [inline chainer/functions/nonparameterized_convolution_2d.py]: I fixed.

File: chainer/functions/nonparameterized_convolution_2d.py (hunk 5 above)
Comment (okuta) [inline chainer/functions/nonparameterized_convolution_2d.py]: I fixed.

File: chainer/functions/nonparameterized_linear.py (hunk 4 above)
Comment (okuta) [inline chainer/functions/nonparameterized_linear.py]: I fixed.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**